src/fileio.c

```diff
@@ -3878,6 +3871,61 @@ vim_rename(char_u *from, char_u *to)
     /*
      * Rename() failed, try copying the file.
      */
+    ret = vim_copyfile(from, to);
+    if (ret != 0)
+	return ret;
+
+    /*
+     * Remove copied original file
+     */
+    if (mch_stat((char *)from, &st) >= 0)
+	mch_remove(from);
+
+    return 0;
+}
+
+
+/*
+ * Create the new file with same permissions as the original.
+ * Return -1 for failure, 0 for success.
+ */
+    int
+vim_copyfile(char_u *from, char_u *to)
+{
+    int		fd_in;
+    int		fd_out;
+    int		n;
+    char	*errmsg = NULL;
+    char	*buffer;
+    long	perm;
+#ifdef HAVE_ACL
+    vim_acl_T	acl;		// ACL from original file
+#endif
+
+#ifdef HAVE_READLINK
+    int		ret;
+    int		len;
+    stat_T	st;
+    char	linkbuf[MAXPATHL + 1];
+
+    ret = mch_lstat((char *)from, &st);
+    if (ret >= 0 && S_ISLNK(st.st_mode))
+    {
+        ret = -1;
+
+	len = readlink((char *)from, linkbuf, MAXPATHL);
+	if (len > 0)
+	{
+	    linkbuf[len] = NUL;
+
+	    // Create link
+	    ret = symlink(linkbuf, (char *)to);
+	}
+
+	return ret == 0 ? ret : -1;
```

Comment:
The function currently returns either 0 or -1.  In this case, is it better to change this function to return a Boolean value (TRUE or FALSE)?

Reply:
No.  Because,

> A detail is that filecopy() may fail for several reasons besides the
file already existing.  If it's not too complicated, the return value
can indicate the reason for why it failed.

Reply:
The current implementation of the function returns only -1 or 0.  This should be modified to return different values based on the error condition.

Reply:
I think it should return OK or FAIL and then please have the caller in `fileio.c` also check for `FAIL/OK`